### PR TITLE
[feat][io] Support configuration secret interpolation

### PR DESCRIPTION
**Pull request by michaeljmarshall**
PIP: https://github.com/apache/pulsar/pull/20903
Relates to: #20862 

### Motivation

The primary motivation is to make it possible to configure Pulsar Connectors in a secure, non-plaintext way. See the PIP for background and relevant details. The new interpolation feature only applies when deploying with functions to Kubernetes.

### Modifications

* Add `SecretsProvider#interpolateSecretForValue` method with a default that maintains the current behavior.
* Override `interpolateSecretForValue` in the `EnvironmentBasedSecretsProvider` so that configuration values formatted as `${my-env-var}` will be replaced with the result of `System.getEnv("my-env-var")` if the result is not `null`.
* Implement a recursive string interpolation method that will replace any configuration value that the `interpolateSecretForValue` implementation determines ought to be replaced.

### Verifying this change

Tests are added/modified.

### Documentation

- [x] `doc-required`

### Matching PR in forked repository

PR in forked repository: https://github.com/michaeljmarshall/pulsar/pull/55

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
+1

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsprovider/EnvironmentBasedSecretsProvider.java [hunk 1]
```diff
@@ -33,4 +41,15 @@ public class EnvironmentBasedSecretsProvider implements SecretsProvider {
     public String provideSecret(String secretName, Object pathToSecret) {
         return System.getenv(secretName);
     }
+
+    @Override
+    public String interpolateSecretForValue(String value) {
```
Comment (shibd) [inline pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsprovider/EnvironmentBasedSecretsProvider.java:46]: It's better to add unit tests to the method separately.

**Review by shibd — APPROVED** [review]
/LGTM

**Pull request merged**